ckanext/security/resource_upload_validator.py

```diff
@@ -0,0 +1,97 @@
+    mimes_instance = mimetypes.MimeTypes()
+    extensions_and_mimetypes = []
+
+    # get supplied file extension and possible mimetypes for that extension
+    _, supplied_file_extension = os.path.splitext(filename)
+    if supplied_file_extension:
+        extensions_and_mimetypes.append(supplied_file_extension)
+
+        guessed_mimetypes = filter(lambda type: type is not None, [
+            mimes_instance.types_map[0].get(supplied_file_extension),
+            mimes_instance.types_map[1].get(supplied_file_extension)
+        ])
+        extensions_and_mimetypes.extend(guessed_mimetypes)
+
+    if file_content:
+        # get inferred mimetype of file and possible extensions for that mimetype
+        try:
+            # try to get python-magic to infer upload file mime type from actual file content
+            mimetype = magic.from_buffer(file_content.read(2048), mime=True)
+        except IOError:
+            log.warning('Unable to detect mimetype from file content')
+
+        if mimetype:
+            extensions_and_mimetypes.append(mimetype)
+
+            # build set of possible extensions for the mimetype
+            nonstandard_extensions = mimes_instance.types_map_inv[0].get(mimetype, [])
+            standard_extensions = mimes_instance.types_map_inv[1].get(mimetype, [])
+            extensions_and_mimetypes.extend(nonstandard_extensions)
+            extensions_and_mimetypes.extend(standard_extensions)
+
+    unique_set = set(extensions_and_mimetypes)
+    unique_list = list(unique_set)
+
+    return unique_list
+
+def validate_upload_type(resource):
+    """
+    Uses the mimetypes builtin library to make inferences about the filename
+    and test the possible mimetypes and extensions against the blacklist.
+    Also uses python-magic to attempt detection of the mimetype by file contents.
+
+    NOTE: When linking files rather than uploading, we only test the extension at present.
+    """
+    try:
+        uploaded_file = resource.get('upload').file
+    except AttributeError:
+        uploaded_file = None
+
+    _add_mimetypes()
+    extensions_and_mimetypes = _build_mimetypes_and_extensions(resource.get('url'), uploaded_file)
+
+    config_blacklist = eval(config.get('ckanext.security.upload_blacklist', '[]'))
+    blacklist = list(DEFAULT_UPLOAD_BLACKLIST)
+    blacklist.extend(config_blacklist)
+
+    log.info('Detected extensions/mimetypes: {}'.format(extensions_and_mimetypes))
+    # test all extensions and mimetypes against blacklist, fail fast
+    if any(map(lambda ext: ext in blacklist, extensions_and_mimetypes)):
+        log.warning('Prevented upload of {}, detected mimetypes/extensions: {}, blacklist: {}'.format(resource.get('url'), extensions_and_mimetypes, blacklist))
+        resource['url'] = ''
+        action = 'upload' if uploaded_file else 'link'
+        raise ValidationError(
+            {'File': ['Cannot {} files of this type'.format(action)]}
+        )
```

Comment:
So does this happen after a file is uploaded, but stops it being linked/processed any further? What happens to the file, is it just left in the temp space and cleaned up by the system eventually?

Reply:
Yeah, the domain object that encapsulates an upload or link is called a Resource, so we're validating the uploaded file before the actual resource gets created. As far as where the uploaded file lives, it's not clear to me from the ckan codebase where the file is at this point. I'll see if I can get any more info about that.

Reply:
As far as I can tell, the resource upload is an instance of cgi.FieldStorage, and this appears to store the file in memory. It's a mapping of an HTML form object on the serverside apparently, making it easier to interact with form data in Python. From here: https://docs.python.org/2/library/cgi.html#using-the-cgi-module
> If a field represents an uploaded file, accessing the value via the value attribute or the getvalue() method reads the entire file in memory as a string.

The resource creation code after the before_create and before_update hooks actually persists the in memory file to disk.